### PR TITLE
Anst/new input system

### LINKED ISSUE #100: Implement user input via Unity's new input system
### Problem or question to be solved

At the moment only keyboard and mouse input is implemented in UltraStar Play.

### Suggested solution/s

In Unity 2020.1 the new input system for Unity will be available in a stable version.

User input should be implemented using the new input system and documented.
I prepared a wiki page for this: https://github.com/UltraStar-Deluxe/Play/wiki/Handling-User-Input.

### PR DESCRIPTION
**Pull request by achimmihca**
### What does this PR do?

This is a major PR that implements all controls using Unity's [new input system](https://docs.unity3d.com/Packages/com.unity.inputsystem@1.0/manual/Actions.html).
It also refines the controls, adds basic Controller support, and implements better touch support.

- Unity's InputActions (IA) are missing a way to resolve conflicts between different actions and different listeners.
    - For example, "Ctrl+E" would also trigger "E".
        - To solve this, one can check that the IA of "Ctrl+E" is not triggered when listening for the IA of "E".
    - Closing a popup dialog via "Escape" could also Exit the scene or application.
        - To solve this, I added a priority-queue where events can be canceled. See ObservableCancelablePriorityInputAction.

- I wanted to subscribe to events via UniRx's reactive API. Thus, Observable...InputAction. 
    - This allows composing events and conditions.
    - This fits well with UniRx observables for controls, e.g., OnClickAsObservable.

**R class**
- The available IA are constants, similar to the available translations. Thus, I extended the constant generation to include available InputAction paths.
    - Example: `R.InputActions.usplay_toggleFavorite == "usplay/toggleFavorite"`
    - I suggest to use camelCase to name IA. This matches our other [property naming conventions](https://github.com/UltraStar-Deluxe/Play/wiki/Translations,-Internationalization-(I18N)).

**InputManager**
- A new InputManager provides static access to IA and manages the currently used InputActions.
    - Example: `InputManager.GetInputAction(R.InputActions.usplay_restartSong).PerformedAsObservable()
            .Subscribe(_ => singSceneController.Restart());`
    - The InputManager is the "owner" of the ObservableCancelablePriorityInputAction. Thus, on destroy of the InputManager (i.e. when the scene is changed), all currently bound InputActions are disposed.
        - Otherwise, disposing of InputActions would be error-prone in my opinion.
        - Stuff that does not listen for an IA from scene-start to scene-end must still explicitly dispose the subscription. I usually do this via a `List<IDisposable> disposables`, to which I add all subscriptions.

**Configurable**
- I wanted the IA to be configurable. The InputActionAsset format is JSON. Thus, the default IA are copied to persistendDataPath and then used from there. Users can access this file and thus override basic bindings.
    - Unity also provides relatively simple API to rebind stuff at runtime. However, this is something for a following PR.

**Controller support in SongSelect**
- The SongSelectScene now implements custom navigation of the controls, mostly for better controller support (see SongSelectSceneControlNavigator).
    - Navigate left/right to select songs. Navigate up/down to select different controls (player profiles, playlist, character-quick-jump, song-order, search)

**Drag to scroll in song select**
- Songs in the SongSelectScene can now be dragged
    - Therefor, I added a "SlotList" with slots, items, and a controlling class.
    - An item can be dragged between its neighboring slots. Other items adjust accordingly, such that the whole list moves.
        - SongRouletteItem implements ISlotListItem. SongRoulettePlaceholder implements ISlotListSlot
        - SongRouletteController holds a SlotListControl that manages the drag events. When an item changes its slot, then one item is removed from one end of the list and another item spawns at the other end of the list.
        - I had to add a new SongRoulettePlaceholder left and right of the screen to ensure that there is always a slot where the item can be dragged to. This changes the animation a little bit.
        - Flick gesture is also supported (I mean continued scrolling if done with "momentum", i.e., remaining mouse-delta / touch-delta).

- The SearchInputField now uses RectTransformSlideIntoViewport, which (although unintuitive) also works on mobile (put your finger there)

**Context menus**
- To make more features available on mobile, I moved more actions into context menus (e.g. "Open song editor").
    - Context menu is opened via right click, but also via long press (1 sec).

**SongEditor**
- Basic editing of songs works also on mobile.
- Several basic editing actions of the song-editor are now available in context menus, e.g., edit lyrics, copy notes, delete notes, ...
- Drag with two fingers to scroll
- Pinch with two fingers to zoom
- Play/Pause via double click. Also added a button for this.

**In-game info of available controls**
- When paused, the SingScene was showing some controls. I extended this to show all currently bound InputActions.
    - This is also available in the SongEditor. Very useful.
    - The InputManager holds a list of AdditionalInputActionInfos. This is used to provide information about other available input handling that is not directly reflected in an dedicated InputAction. For example "Navigate Right" (left/right/up/down) is used to "skip to the next lyrics" in the SingScene.

- I removed other Icons from the scene that were showing available inputs (e.g. arrow keys to select highscore).
    - I had only added these because I saw them in USDX.
    - Instead, I would now suggest to improve the dynamically generated InputActionInfos.
    - Also, I added a button to iterate the highscores (mostly for mobile)

**Small improvements to controls**
- "Toggle pause" in sing scene via double click
- Fuzzy search is now performed in the SongSelect scene by default, i.e., when no modifier key is pressed.
    - Thus, "Open song editor" is now "Ctrl+E" instead of just "E"
- Toggle search has been changed to "Ctrl+F" because this is a super common shortcut.
- To improve controller support, I removed unused Button components from sliders. There was always a Button behind the value label.

### Closes Issue(s)

Close #100
Also, implements most of the stuff from #184

### More

- Controls have changed a little. For example, use Ctrl+E instead of just E to open the editor.
- Will update documentation shortly
- Overall I am very happy with this new input system.

### Additional Notes

- Basic understanding of the new input system is needed
    - An InputAction (e.g. "copy" or "open editor") can be associated with several bindings (e.g. "Ctrl+C", "Ctrl+E", double click)
    - An InputAction comes with a value. For a Controller stick or mouse this is a Vector2. For Buttons it is a float (0 not pressed, 1 pressed). Multiple buttons together can act as one more complex input for an InputActions (e.g. arrow keys together provide a Vector2)
    - For Controllers the general button-layout is called ButtonNorth, ButtonWest, etc.

- For review, I suggest to start with the InputManager, and ObservableCancelablePriorityInputAction.
    - Also start with a class that uses the resulting API, e.g., SingSceneInputControl

- Because they are not limited to keyboard input anymore, the classes are now called `...InputControl` instead of `...KeyboardInputController`

- Next I want to implement #217 because this is relevant for every new user. Afterwards, I will create a release.

### CODE REVIEW

**Comment by achimmihca** [comment]
The remaining Codacy issues are on purpose.

**Comment by achimmihca** [comment]
rebased onto master

File: UltraStar Play/Assets/Scenes/SongSelect/SongRoulette/SongRouletteController.cs [hunk 1]
```diff
@@ -100,13 +205,13 @@ private void UpdateSongRouletteItems()
             SongRouletteItem songRouletteItem = songRouletteItems.Where(it => it.SongMeta == songMeta).FirstOrDefault();
             if (songRouletteItem == null)
             {
-                songRouletteItem = CreateSongRouletteItem(songMeta, rouletteItem);
+                songRouletteItem = CreateSongRouletteItem(songMeta, placeholder);
             }
 
             // Update target
-            if (songRouletteItem.TargetRouletteItem != rouletteItem)
+            if (songRouletteItem.TargetRouletteItem != placeholder)
```
[the remark this answers is not in the document]
Comment (achimmihca) [reply UltraStar Play/Assets/Scenes/SongSelect/SongRoulette/SongRouletteController.cs]: Moved the is-value-differnt check to the setter/property. Makes more sense now.

File: UltraStar Play/Assets/Common/Input/GlobalInputControl.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+            .Subscribe(_ => AudioManager.ToggleMuteAudio());
+    }
+
+    private void ToggleFullscreen()
+    {
+        Debug.Log("Toggle full-screen mode");
+        Screen.fullScreen = !Screen.fullScreen;
+        // Screen.fullScreenMode is updated after this frame. Thus, delay updating the settings a bit.
+        StartCoroutine(CoroutineUtils.ExecuteAfterDelayInSeconds(0.1f,
+            () => SettingsManager.Instance.Settings.GraphicSettings.fullScreenMode = Screen.fullScreenMode));
```
[the remark this answers is not in the document]
Comment (achimmihca) [reply UltraStar Play/Assets/Common/Input/GlobalInputControl.cs]: See https://docs.unity3d.com/ScriptReference/Screen-fullScreen.html
The delay is one frame. I changed it to update the settings value after two frames, which should work on all systems.

**Review by basisbit — APPROVED** [review]

**Comment by basisbit** [comment]
good work! Thank you!

**Pull request merged**